### PR TITLE
feat: add canonical cache for live sync

### PR DESCRIPTION
**Pull request by forcodedancing**
### Description

Add cache layer to cache historical committed accounts & storages.

### Rationale

To improve the execution of live sync.

#### Test Result - Live sync 20K blocks on BSC, from 41413200 to 41433100.

Note: There are reboots before each tests, to make sure the test environment is clean.

* Metrics 1 - hit rate
image

account hit rate (the l3 account in the figure): 36.2%

storage hit rate (the l3 storage in the figure): 43.3%

Note: l1 account & storage caches refer to the revm state cache.

* Metrics 2 - the execution time for every 100 blocks

image

image

The time decreases about 21% after steady. (due to reboot, at the beginning, the performance is bad for both). 

The time to apply bundle state to cache (update cache after db commit) is included for the cached version.

* Metrics 3 - the gas throughput for [41432701, 41433100]

image

The cached version is 356 Mgas/s for execution on avg, while the original version is 277 Mgas/s.

The gas throughput increases about 28%.  Why it is not aligned to the execution time?
- The avg Mgas/s is calculated from the log (each block has a log about it).
- The cached version introduces some extra cost for applying state to cache. This extra time is included in the Metrics 2.

### Changes

Notable changes: 
* add each change in a bullet point here
* ...

### Potential Impacts
* add potential impacts for other components here
* ...

### CODE REVIEW

File: crates/blockchain-tree/src/canonical_cache.rs [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+                    },
+                );
+            }
+        }
+    }
+
+    let mut to_wipe = false;
+    for storage in &change_set.storage {
+        if storage.wipe_storage {
+            to_wipe = true;
```
Comment (unclezoro) [inline crates/blockchain-tree/src/canonical_cache.rs:57]: `storage.wipe_storage` seems only mean to the storage of an account, but why we clean all storage cache here?

File: crates/blockchain-tree/src/canonical_cache.rs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+const CACHE_SIZE: usize = 1000000;
+
+type AddressStorageKey = (Address, StorageKey);
+
+lazy_static! {
+    /// Account cache
+    pub static ref ACCOUNT_CACHE: Cache<Address, Account> = Cache::new(CACHE_SIZE);
+
+    /// Contract cache
+    static ref CONTRACT_CACHE: Cache<B256, Bytecode> = Cache::new(CACHE_SIZE);
```
Comment (unclezoro) [inline crates/blockchain-tree/src/canonical_cache.rs]: I think this is too big, because the smart contract code is long, and the hot contract is less than `1000000`

File: crates/blockchain-tree/src/canonical_cache.rs [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        }
+        Ok(None)
+    }
+
+    fn canonical_hashes_range(
+        &self,
+        _start: BlockNumber,
+        _end: BlockNumber,
+    ) -> ProviderResult<Vec<B256>> {
+        unimplemented!()
```
Comment (unclezoro) [inline crates/blockchain-tree/src/canonical_cache.rs:114]: is this ok?

File: crates/blockchain-tree/src/canonical_cache.rs [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        if let Some(value) = self.state_provider.basic_account(address)? {
+            ACCOUNT_CACHE.insert(address, value);
+            return Ok(Some(value))
+        }
+        Ok(None)
+    }
+}
+
+impl<SP: StateProvider, EDP: ExecutionDataProvider> StateRootProvider
+    for CachedBundleStateProvider<SP, EDP>
```
Comment (unclezoro) [inline crates/blockchain-tree/src/canonical_cache.rs:139]: this part seems not related with cache, and it is similar to `BundleStateProvider`, can we reduce the duplicated code in some way?

File: crates/blockchain-tree/src/blockchain_tree.rs [hunk 5]
```diff
@@ -1261,6 +1262,7 @@ where
         };
         recorder.record_relative(MakeCanonicalAction::RetrieveStateTrieUpdates);
 
+        let cloned_bundle = state.clone().bundle;
```
Comment (unclezoro) [inline crates/blockchain-tree/src/blockchain_tree.rs]: --> `state.bundle.clone()`?

File: crates/blockchain-tree/src/canonical_cache.rs (hunk 1 above)
Comment (forcodedancing) [reply crates/blockchain-tree/src/canonical_cache.rs:57]: The reason is that: the quick cache is very simple there is no function for iterate or find keys with a prefix. 
One way is that we can make storage cache to use nested structure, e.g., a map or another layer of cache. However, it seems a bit complex. Do you think should we implement in this way?

File: crates/blockchain-tree/src/canonical_cache.rs (hunk 2 above)
Comment (forcodedancing) [reply crates/blockchain-tree/src/canonical_cache.rs]: Sure. Let me change it.

File: crates/blockchain-tree/src/canonical_cache.rs (hunk 3 above)
Comment (forcodedancing) [reply crates/blockchain-tree/src/canonical_cache.rs:114]: Yes. It is ok. The `BundleStateProvider` also does not implement it.

File: crates/blockchain-tree/src/blockchain_tree.rs (hunk 5 above)
Comment (forcodedancing) [reply crates/blockchain-tree/src/blockchain_tree.rs]: fixed

File: crates/blockchain-tree/src/canonical_cache.rs (hunk 4 above)
Comment (forcodedancing) [reply crates/blockchain-tree/src/canonical_cache.rs:139]: Have tired, however it is not easy, for `StateProvider` and `ExecutionDataProvider` are not cleanable - cannot  construct a  `BundleStateProvider` inside to reduce codes.

File: crates/blockchain-tree/src/canonical_cache.rs (hunk 1 above)
Comment (unclezoro) [reply crates/blockchain-tree/src/canonical_cache.rs:57]: I think we should have some test, like create/destroy a contract to ensure our cache works well
